### PR TITLE
feat(unstable): add OTEL MeterProvider

### PR DESCRIPTION
**Pull request by lucacasonato**
This commit replaces `Deno.telemetry.MetricsExporter` with `Deno.telemetry.MeterProvider`.

### CODE REVIEW

**Comment by lucacasonato** [comment]
need to wire up `OTEL_EXPORTER_OTLP_METRICS_TEMPORALITY_PREFERENCE` and `OTEL_EXPORTER_OTLP_METRICS_DEFAULT_HISTOGRAM_AGGREGATION` before landing

**Comment by bartlomieju** [comment]
@lucacasonato can you change the name to `feat(unstable): ...` so we can ship it in a patch?

File: ext/telemetry/telemetry.ts [hunk 1]
```diff
@@ -688,260 +696,482 @@ class ContextManager {
...
+    recordObservable(this.#instrument, value, attributes);
+  }
+}
+
+async function observe(): Promise<void> {
+  const promises: Promise<void>[] = [];
+  for (const [observable, callbacks] of INDIVIDUAL_CALLBACKS) {
+    const result = getObservableResult(observable);
+    for (const callback of callbacks) {
+      promises.push(Promise.try(() => callback(result)));
```
Comment (devsnek) [inline ext/telemetry/telemetry.ts:1151]:
```suggestion
      promises.push(Promise.try(callback, result));
```

File: ext/telemetry/lib.rs [hunk 2]
```diff
@@ -976,538 +1117,579 @@ fn op_otel_span_flush(state: &mut OpState) {
...
+    None => match (attr1, attr2, attr3) {
+      (Some(kv1), Some(kv2), Some(kv3)) => &[kv1, kv2, kv3] as &[KeyValue],
+      (Some(kv1), Some(kv2), None) => &[kv1, kv2],
+      (Some(kv1), None, Some(kv3)) => &[kv1, kv3],
+      (None, Some(kv2), Some(kv3)) => &[kv2, kv3],
+      (Some(kv1), None, None) => &[kv1],
+      (None, Some(kv2), None) => &[kv2],
+      (None, None, Some(kv3)) => &[kv3],
+      (None, None, None) => &[],
+    },
```
[the remark this answers is not in the document]
Comment (lucacasonato) [reply ext/telemetry/lib.rs:1540]: but then i need to collect it into a vec (allocations!)

File: ext/telemetry/telemetry.ts [hunk 3]
```diff
@@ -688,260 +696,482 @@ class ContextManager {
...
+let isObserving = false;
+function startObserving() {
+  if (!isObserving) {
+    isObserving = true;
+    (async () => {
+      while (true) {
+        const promise = op_otel_metric_wait_to_observe();
+        core.unrefOpPromise(promise);
+        const ok = await promise;
+        if (!ok) break;
```
[the remark this answers is not in the document]
Comment (lucacasonato) [reply ext/telemetry/telemetry.ts:1169]: yes - should never happen I think, but we handle it anyway

File: ext/telemetry/telemetry.ts [hunk 4]
```diff
@@ -687,260 +700,515 @@ class ContextManager {
...
+  addCallback(_callback: ObservableCallback) {}
+  removeCallback(_callback: ObservableCallback) {}
+}
+
+type BatchObservableCallback = (
+  observableResult: BatchObservableResult,
+) => void | Promise<void>;
+
+function record(
+  instrument: Instrument,
```
Comment (devsnek) [inline ext/telemetry/telemetry.ts:947]: if this accepts `Instrument|null` you can pipe that through the normal classes based on `METRICS_ENABLED` without needing a NoopInstrument class (smth like `let instrument = METRICS_ENABLED ? op_create_instrument() : null`)

**Review by devsnek — APPROVED** [review]

**Pull request merged**